store/sql_reaction_store.go

```diff
@@ -0,0 +1,239 @@
+
+type SqlReactionStore struct {
+	*SqlStore
+}
+
+func NewSqlReactionStore(sqlStore *SqlStore) ReactionStore {
+	s := &SqlReactionStore{sqlStore}
+
+	for _, db := range sqlStore.GetAllConns() {
+		table := db.AddTableWithName(model.Reaction{}, "Reactions").SetKeys(false, "UserId", "PostId", "EmojiName")
+		table.ColMap("UserId").SetMaxSize(26)
+		table.ColMap("PostId").SetMaxSize(26)
+		table.ColMap("EmojiName").SetMaxSize(64)
+	}
+
+	return s
+}
+
+func (s SqlReactionStore) CreateIndexesIfNotExists() {
+	s.CreateIndexIfNotExists("idx_reactions_post_id", "Reactions", "PostId")
+}
+
+func (s SqlReactionStore) Save(reaction *model.Reaction) StoreChannel {
+	storeChannel := make(StoreChannel)
+
+	go func() {
+		result := StoreResult{}
+
+		reaction.PreSave()
+		if result.Err = reaction.IsValid(); result.Err != nil {
+			storeChannel <- result
+			close(storeChannel)
+			return
+		}
+
+		if transaction, err := s.GetMaster().Begin(); err != nil {
+			result.Err = model.NewLocAppError("SqlReactionStore.Save", "store.sql_reaction.save.begin.app_error", nil, err.Error())
+		} else {
+			err := saveReactionAndUpdatePost(transaction, reaction)
+
+			if err != nil {
+				transaction.Rollback()
+
+				result.Err = model.NewLocAppError("SqlPreferenceStore.Save", "store.sql_reaction.save.save.app_error", nil, err.Error())
+			} else if err := transaction.Commit(); err != nil {
+				// don't need to rollback here since the transaction is already closed
+				result.Err = model.NewLocAppError("SqlPreferenceStore.Save", "store.sql_preference.save.commit.app_error", nil, err.Error())
+			} else {
+				result.Data = reaction
+			}
+		}
+
+		storeChannel <- result
+		close(storeChannel)
+	}()
+
+	return storeChannel
+}
+
+func (s SqlReactionStore) Delete(reaction *model.Reaction) StoreChannel {
+	storeChannel := make(StoreChannel)
+
+	go func() {
+		result := StoreResult{}
+
+		if transaction, err := s.GetMaster().Begin(); err != nil {
+			result.Err = model.NewLocAppError("SqlReactionStore.Save", "store.sql_reaction.delete.begin.app_error", nil, err.Error())
+		} else {
+			err := deleteReactionAndUpdatePost(transaction, reaction)
+
+			if err != nil {
+				transaction.Rollback()
+
+				result.Err = model.NewLocAppError("SqlPreferenceStore.Save", "store.sql_reaction.delete.app_error", nil, err.Error())
+			} else if err := transaction.Commit(); err != nil {
+				// don't need to rollback here since the transaction is already closed
+				result.Err = model.NewLocAppError("SqlPreferenceStore.Save", "store.sql_preference.delete.commit.app_error", nil, err.Error())
+			} else {
+				result.Data = reaction
+			}
+		}
+
+		storeChannel <- result
+		close(storeChannel)
+	}()
+
+	return storeChannel
+}
+
+func saveReactionAndUpdatePost(transaction *gorp.Transaction, reaction *model.Reaction) error {
+	if count, err := transaction.SelectInt(
+		`SELECT
+			COUNT(0)
+		FROM
+			Reactions
+		WHERE
+			PostId = :PostId AND
+			UserId = :UserId AND
+			EmojiName = :EmojiName`,
```

Comment:
Correct me if im wrong, but if this table has as PK `UserId`, `PostId` and `EmojiName` this check is redundant as the insert should fail with a duplicate key error